File: Day5/Part1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// Read and split input
	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var content []string
	for scanner.Scan() {
		content = append(content, strings.TrimSpace(scanner.Text()))
	}

	splitIndex := 0
	for i, line := range content {
		if line == "" {
			splitIndex = i
			break
		}
	}

	rules := parseRules(content[:splitIndex])
	updates := parseUpdates(content[splitIndex+1:])

	nPart1 := 0

	// Check if all rules are correct for each update
	for _, values := range updates {
		relevantRules := filterRelevantRules(rules, values)
		if allCorrect(values, relevantRules) {
			midIndex := len(values) / 2
			val := atoi(values[midIndex])
			nPart1 += val
		}
	}

	fmt.Printf("%d\n", nPart1)
}

func parseRules(lines []string) [][2]string {
	var rules [][2]string
	for _, line := range lines {
		parts := strings.Split(line, "|")
		rules = append(rules, [2]string{parts[0], parts[1]})
	}
	return rules
}

func parseUpdates(lines []string) [][]string {
	var updates [][]string
	for _, line := range lines {
		updates = append(updates, strings.Split(line, ","))
	}
	return updates
}

func filterRelevantRules(rules [][2]string, values []string) [][2]string {
	var relevantRules [][2]string
	for _, rule := range rules {
		if contains(values, rule[0]) && contains(values, rule[1]) {
			relevantRules = append(relevantRules, rule)
		}
	}
	return relevantRules
}

func allCorrect(values []string, rules [][2]string) bool {
	for _, rule := range rules {
		if indexOf(values, rule[0]) >= indexOf(values, rule[1]) {
			return false
		}
	}
	return true
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func indexOf(slice []string, item string) int {
	for i, v := range slice {
		if v == item {
			return i
		}
	}
	return -1
}

func atoi(str string) int {
	num, _ := strconv.Atoi(str)
	return num
}

File: Day5/Part1/main.py
def main():
    # Read and split input
    with open("input.txt") as file:
        content = [line.strip() for line in file]

    # Find split index
    split_index = content.index("")

    rules = parse_rules(content[:split_index])
    updates = parse_updates(content[split_index + 1:])

    n_part1 = 0

    # Check if all rules are correct for each update
    for values in updates:
        relevant_rules = filter_relevant_rules(rules, values)
        if all_correct(values, relevant_rules):
            mid_index = len(values) // 2
            val = atoi(values[mid_index])
            n_part1 += val

    print(n_part1)


def parse_rules(lines):
    rules = []
    for line in lines:
        parts = line.split("|")
        rules.append((parts[0], parts[1]))
    return rules


def parse_updates(lines):
    updates = []
    for line in lines:
        updates.append(line.split(","))
    return updates


def filter_relevant_rules(rules, values):
    relevant_rules = []
    for rule in rules:
        if contains(values, rule[0]) and contains(values, rule[1]):
            relevant_rules.append(rule)
    return relevant_rules


def all_correct(values, rules):
    for rule in rules:
        if index_of(values, rule[0]) >= index_of(values, rule[1]):
            return False
    return True


def contains(lst, item):
    return item in lst


def index_of(lst, item):
    try:
        return lst.index(item)
    except ValueError:
        return -1


def atoi(s):
    try:
        return int(s)
    except ValueError:
        return 0


if __name__ == "__main__":
    main()
